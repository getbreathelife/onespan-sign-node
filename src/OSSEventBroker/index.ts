export * from './OSSEventBroker';
export * from './error';
