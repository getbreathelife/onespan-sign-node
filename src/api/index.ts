import * as RequestBuilders from './requestBuilders';

export { Api } from './Api';
export { RequestBuilders };
