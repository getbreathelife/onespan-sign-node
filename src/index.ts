/**
 * A library to interact with OneSpan Sign's API.
 *
 * @example
 * ```ts
 * const oneSpanSign = new OneSpanSign({
 *   clientId: CLIENT_ID,
 *   secret: CLIENT_SECRET,
 *   type: 'OWNER',   // or 'SENDER'
 * });
 *
 * const { id: packageId } = await oneSpanSign.packages.create({...});
 *
 * await oneSpanSign.documents.create(packageId, ...)
 * ```
 *
 * @packageDocumentation
 */

export { OneSpanSign } from './OneSpanSign';
export * from './api';
export * from './resources';
export * from './types';
export * from './types/requests';
