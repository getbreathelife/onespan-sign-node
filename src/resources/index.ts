export { Resource } from './Resource';
export { DocumentResource } from './DocumentResource';
export { PackageResource } from './PackageResource';
