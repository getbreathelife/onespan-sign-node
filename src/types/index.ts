// Entity typing reference: https://community.onespan.com/sites/default/files/interactive_api_references/oss_1145.yaml
import * as Events from './events';
import * as Requests from './requests';
import * as Responses from './responses';

export * from './api';
export * from './models';
export * from './utils';
export { Requests, Responses, Events };
