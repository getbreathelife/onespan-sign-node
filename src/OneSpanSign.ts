import FormData from 'form-data';

import { Api } from './api';
import {
  CreatePackageRequestPayload,
  CreatePackageResponsePayload,
  DocumentMetadata,
  UploadDocumentRequestPayload,
} from './types';

export class OneSpanSign {
  constructor(private apiKey: string, private apiUrl: string) {}

  public async createPackage(payload: CreatePackageRequestPayload): Promise<CreatePackageResponsePayload> {
    const response = await Api.post(`${this.apiUrl}/api/packages`)
      .withAuthorizationHeader(`Basic ${this.apiKey}`)
      .withBody(payload)
      .fetch();

    return (await response.json()) as CreatePackageResponsePayload;
  }

  public async uploadDocument(
    packageId: string,
    payload: UploadDocumentRequestPayload,
    documentBody: Buffer | ReadableStream
  ): Promise<DocumentMetadata> {
    const formData = new FormData();
    formData.append('file', documentBody, { filename: payload.name });
    formData.append('payload', payload);

    const response = await Api.post(`${this.apiUrl}/api/packages/${packageId}/documents`)
      .withAuthorizationHeader(`Basic ${this.apiKey}`)
      .withBody(formData)
      .fetch();

    return (await response.json()) as DocumentMetadata;
  }
}
