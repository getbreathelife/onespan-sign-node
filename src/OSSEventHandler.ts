import { EventHandler, Events } from 'types';

/**
 * Helper class to ingest
 * {@link https://community.onespan.com/documentation/onespan-sign/guides/feature-guides/developer/setting-callback-notifications | callback notifications}
 * from OneSpan Sign and call the appropriate handler.
 *
 * @public
 */
export class OSSEventBroker {
  protected _documentSignedHandler?: EventHandler<Events.DocumentSignedEvent>;

  public get documentSignedHandler(): EventHandler<Events.DocumentSignedEvent> | undefined {
    return this._documentSignedHandler;
  }

  public set documentSignedHandler(handler) {
    this._documentSignedHandler = handler;
  }
}
