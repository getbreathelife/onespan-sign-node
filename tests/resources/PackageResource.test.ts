import { PackageResource, Requests, Api } from '../../src';
import { mockedFetch, mockFetchHappyPath } from '../setup/mockNodeFetch';

const MOCK_API_KEY = 'demoKey';
const MOCK_API_URL = 'http://demo.com';

describe('PackageResource', () => {
  let packages: PackageResource;

  beforeAll(() => {
    mockFetchHappyPath();
    packages = new PackageResource(new Api(MOCK_API_KEY, MOCK_API_URL));
  });

  afterAll(() => {
    mockedFetch.mockReset();
  });

  afterEach(() => {
    mockedFetch.mockClear();
  });

  describe('create', () => {
    it('sends the provided payload as request body', async () => {
      const payload: Requests.CreatePackageData = {
        name: 'mock package',
      };

      await packages.create(payload);

      expect(mockedFetch.mock.calls[0]).toMatchSnapshot();
    });
  });

  describe('getAll', () => {
    it('sends the provided parameters as query parameters', async () => {
      const params: Requests.GetAllPackagesParameters = {
        query: 'inbox',
        predefined: 'all',
        sort: undefined,
      };

      await packages.getAll(params);

      expect(mockedFetch.mock.calls[0]).toMatchSnapshot();
    });

    it('does not append any query parameter if no parameter is passed in', async () => {
      await packages.getAll();

      expect(mockedFetch.mock.calls[0]).toMatchSnapshot();
    });

    it('serializes Date objects into ISO strings', async () => {
      const params: Requests.GetAllPackagesParameters = {
        query: 'inbox',
        predefined: 'all',
        sort: undefined,
        lastUpdatedStartDate: new Date(1641852302242),
        lastUpdatedEndDate: new Date(1641852302242),
      };

      await packages.getAll(params);

      expect(mockedFetch.mock.calls[0]).toMatchSnapshot();
    });
  });

  describe('getOne', () => {
    it('sends request to the correct URL', async () => {
      await packages.getOne('package1');
      expect(mockedFetch.mock.calls[0]).toMatchSnapshot();
    });
  });

  describe('update', () => {
    it('sends the provided payload as request body', async () => {
      const payload: Requests.UpdatePackageData = {
        name: 'mock package update',
        status: 'ARCHIVED',
      };

      await packages.update('package1', payload);

      expect(mockedFetch.mock.calls[0]).toMatchSnapshot();
    });
  });

  describe('delete', () => {
    it('sends request to the correct URL', async () => {
      await packages.delete('package1');
      expect(mockedFetch.mock.calls[0]).toMatchSnapshot();
    });
  });

  describe('getAuditTrail', () => {
    it('sends request to the correct URL', async () => {
      await packages.getAuditTrail('package1');
      expect(mockedFetch.mock.calls[0]).toMatchSnapshot();
    });
  });

  describe('getEvidenceSummary', () => {
    it('sends request to the correct URL', async () => {
      await packages.getEvidenceSummary('package1');
      expect(mockedFetch.mock.calls[0]).toMatchSnapshot();
    });
  });
});
