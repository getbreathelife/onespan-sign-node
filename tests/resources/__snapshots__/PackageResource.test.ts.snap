// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PackageResource create sends the provided payload as request body 1`] = `
Array [
  "http://demo.com/api/packages",
  Object {
    "body": "{\\"name\\":\\"mock package\\"}",
    "headers": Object {
      "accept": "application/json",
      "authorization": "Basic demoKey",
      "content-type": "application/json",
    },
    "method": "POST",
  },
]
`;

exports[`PackageResource delete sends request to the correct URL 1`] = `
Array [
  "http://demo.com/api/packages/package1",
  Object {
    "headers": Object {
      "accept": "application/json",
      "authorization": "Basic demoKey",
      "content-type": "application/json",
    },
    "method": "DELETE",
  },
]
`;

exports[`PackageResource getAll does not append any query parameter if no parameter is passed in 1`] = `
Array [
  "http://demo.com/api/packages",
  Object {
    "headers": Object {
      "accept": "application/json",
      "authorization": "Basic demoKey",
      "content-type": "application/json",
    },
    "method": "GET",
  },
]
`;

exports[`PackageResource getAll sends the provided parameters as query parameters 1`] = `
Array [
  "http://demo.com/api/packages?query=inbox&predefined=all",
  Object {
    "headers": Object {
      "accept": "application/json",
      "authorization": "Basic demoKey",
      "content-type": "application/json",
    },
    "method": "GET",
  },
]
`;

exports[`PackageResource getAll serializes Date objects into ISO strings 1`] = `
Array [
  "http://demo.com/api/packages?query=inbox&predefined=all&lastUpdatedStartDate=2022-01-10T22%3A05%3A02.242Z&lastUpdatedEndDate=2022-01-10T22%3A05%3A02.242Z",
  Object {
    "headers": Object {
      "accept": "application/json",
      "authorization": "Basic demoKey",
      "content-type": "application/json",
    },
    "method": "GET",
  },
]
`;

exports[`PackageResource getAuditTrail sends request to the correct URL 1`] = `
Array [
  "http://demo.com/api/packages/package1/audit",
  Object {
    "headers": Object {
      "accept": "application/json",
      "authorization": "Basic demoKey",
      "content-type": "application/json",
    },
    "method": "GET",
  },
]
`;

exports[`PackageResource getEvidenceSummary sends request to the correct URL 1`] = `
Array [
  "http://demo.com/api/packages/package1/evidence/summary",
  Object {
    "headers": Object {
      "accept": "application/pdf",
      "authorization": "Basic demoKey",
      "content-type": "application/json",
    },
    "method": "GET",
  },
]
`;

exports[`PackageResource getOne sends request to the correct URL 1`] = `
Array [
  "http://demo.com/api/packages/package1",
  Object {
    "headers": Object {
      "accept": "application/json",
      "authorization": "Basic demoKey",
      "content-type": "application/json",
    },
    "method": "GET",
  },
]
`;

exports[`PackageResource update sends the provided payload as request body 1`] = `
Array [
  "http://demo.com/api/packages/package1",
  Object {
    "body": "{\\"name\\":\\"mock package update\\",\\"status\\":\\"ARCHIVED\\"}",
    "headers": Object {
      "accept": "application/json",
      "authorization": "Basic demoKey",
      "content-type": "application/json",
    },
    "method": "PUT",
  },
]
`;
